crates/core_simd/src/vector.rs

```diff
@@ -974,3 +1119,44 @@ where
 {
     type Mask = isize;
 }
+
+const USE_BRANCH: bool = false;
+const USE_BITMASK: bool = false;
+
+#[inline]
+fn index<T, const N: usize>() -> Simd<T, N>
+where
+    T: MaskElement + Default + core::convert::From<i8> + core::ops::Add<T, Output = T>,
+    LaneCount<N>: SupportedLaneCount,
+{
+    let mut index = [T::default(); N];
+    for i in 1..N {
+        index[i] = index[i - 1] + T::from(1);
+    }
+    Simd::from_array(index)
+}
+
+#[inline]
+fn mask_up_to<M, const N: usize>(enable: Mask<M, N>, len: usize) -> Mask<M, N>
+where
+    LaneCount<N>: SupportedLaneCount,
+    M: MaskElement + Default + core::convert::From<i8> + core::ops::Add<M, Output = M>,
+    Simd<M, N>: SimdPartialOrd,
+    // <Simd<M, N> as SimdPartialEq>::Mask: Mask<M, N>,
+    Mask<M, N>: core::ops::BitAnd<Output = Mask<M, N>> + core::convert::From<Mask<i8, N>>,
+{
+    let index = index::<i8, N>();
+    enable & Mask::<M, N>::from(index.simd_lt(Simd::splat(i8::try_from(len).unwrap_or(i8::MAX))))
+}
+
+// This function matches the semantics of the `bzhi` instruction on x86 BMI2
+// TODO: optimize it further if possible
+// https://stackoverflow.com/questions/75179720/how-to-get-rust-compiler-to-emit-bzhi-instruction-without-resorting-to-platform
+#[inline(always)]
+fn bzhi_u64(a: u64, ix: u32) -> u64 {
```

Comment:
If we made `ix` a `u64`, this would remove the need to do `std::cmp::min(N, slice.len() as u32)` in call-sites above.

Which is doubly important if we take it further:
```rust

extern "C" {
    #[link_name = "llvm.x86.bmi.bzhi.64"]
    fn x86_bmi2_bzhi_64(x: u64, y: u64) -> u64;
}

#[inline(always)]
fn bzhi_u64(a: u64, ix: u64) -> u64 {
    #[cfg(target_feature = "bmi2")]
    unsafe {
        return x86_bmi2_bzhi_64(a, ix);
    }

    if ix > 63 {
        a
    } else {
        a & (1u64 << ix) - 1
    }
}
```

For whatever reason the intel intrinsic of `_bzhi_u64` takes a `ix: u32`, also in the C/C++ versions. Even a `1u64 as u32` generates a redundant register mov to a 32-bit register, clearing the high bits. All of this is truly unnecessary because the instruction operates on 64b registers.

It's the difference between

```asm
        cmp rsi, 64
        mov r9d, 64
        cmovb r9, rsi
        bzhi r9, r8, r9

```
In the current pure-Rust version,

 ```asm
        mov r9d, esi
        bzhi r9, rdx, r9
```
When calling the `core::arch::x86_64::_bzhi_u64` intrinsic with `slice.len() as u32` (which is incorrect for slices longer than u32::MAX).

```asm
        bzhi r9, rdx, rsi
```

When you use the above code with `#![feature(llvm_link_intrinsics)]`

The layers of mismatched semantics are working against us

Reply:
Nevermind, the version without `min(N, slice.len())` is wrong because the instruction only looks at the lowest 8 bits. So it effectively does a modulo 256 on the second operand.

So `x86_bmi2_bzhi_64(u64::MAX, 525)` is equal to `0x1FFF`, 13 low bits set, because `525 % 256 = 13`